// @commercetools-frontend/application-shell/dist/redirect-to-project-create-c962f72f.cjs.prod.js
'use strict';

var react = require('react');
var applicationComponents = require('@commercetools-frontend/application-components');
var notifications = require('@commercetools-uikit/notifications');
var Spacings = require('@commercetools-uikit/spacings');
var Text = require('@commercetools-uikit/text');
var useIsServedByProxy = require('./use-is-served-by-proxy-845cace5.cjs.prod.js');
var getMcOrigin = require('./get-mc-origin-8f7a578c.cjs.prod.js');
var jsxRuntime = require('@emotion/react/jsx-runtime');
require('@commercetools-frontend/application-shell-connectors');
require('@babel/runtime-corejs3/helpers/slicedToArray');
require('@babel/runtime-corejs3/core-js-stable/url');
require('@babel/runtime-corejs3/core-js-stable/instance/reverse');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var Spacings__default = /*#__PURE__*/_interopDefault(Spacings);
var Text__default = /*#__PURE__*/_interopDefault(Text);

const RedirectToProjectCreate = () => {
  const servedByProxy = useIsServedByProxy.useIsServedByProxy();
  react.useEffect(() => {
    if (servedByProxy === true) {
      useIsServedByProxy.location.replace('/account/projects/new');
      return;
    }
  }, [servedByProxy]);
  if (servedByProxy) {
    return null;
  }
  const mcUrl = getMcOrigin.getMcOrigin(window.app.mcApiUrl);
  return jsxRuntime.jsx("div", {
    children: jsxRuntime.jsx(Spacings__default["default"].Inset, {
      scale: "xl",
      children: jsxRuntime.jsx(applicationComponents.PageContentNarrow, {
        children: jsxRuntime.jsxs(Spacings__default["default"].Stack, {
          children: [jsxRuntime.jsx(Text__default["default"].Headline, {
            as: "h1",
            children: "Please create a project!"
          }), jsxRuntime.jsx(notifications.ContentNotification, {
            type: "warning",
            children: "You are running in development mode"
          }), jsxRuntime.jsxs(Text__default["default"].Body, {
            children: ["The application is not running behind the", ' ', jsxRuntime.jsx("a", {
              href: "https://docs.commercetools.com/custom-applications/concepts/merchant-center-proxy-router",
              target: "_blank",
              rel: "noopener noreferrer",
              children: "Merchant Center Proxy"
            }), ' ', "in production and thus cannot be redirected to the account section to create a new project."]
          }), jsxRuntime.jsxs(Text__default["default"].Body, {
            children: ["Instead, we recommend to go to the", ' ', jsxRuntime.jsx("a", {
              href: mcUrl,
              target: "_blank",
              rel: "noopener noreferrer",
              children: "Merchant Center production URL"
            }), ' ', "and create a project there. After that, you can access your new project from your local environment."]
          })]
        })
      })
    })
  });
};
RedirectToProjectCreate.displayName = 'RedirectToProjectCreate';

exports.RedirectToProjectCreate = RedirectToProjectCreate;
exports["default"] = RedirectToProjectCreate;
