// @commercetools-uikit/notifications/dist/commercetools-uikit-notifications.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _Object$keys = require('@babel/runtime-corejs3/core-js-stable/object/keys');
var _Object$getOwnPropertySymbols = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
var _filterInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/filter');
var _Object$getOwnPropertyDescriptor = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
var _forEachInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
var _Object$getOwnPropertyDescriptors = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
var _Object$defineProperties = require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
var _Object$defineProperty = require('@babel/runtime-corejs3/core-js-stable/object/define-property');
var _defineProperty = require('@babel/runtime-corejs3/helpers/defineProperty');
require('prop-types');
var react$1 = require('react');
var reactIntl = require('react-intl');
var react = require('@emotion/react');
var AccessibleButton = require('@commercetools-uikit/accessible-button');
var designSystem = require('@commercetools-uikit/design-system');
var icons = require('@commercetools-uikit/icons');
var utils = require('@commercetools-uikit/utils');
var jsxRuntime = require('@emotion/react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _Object$keys__default = /*#__PURE__*/_interopDefault(_Object$keys);
var _Object$getOwnPropertySymbols__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertySymbols);
var _filterInstanceProperty__default = /*#__PURE__*/_interopDefault(_filterInstanceProperty);
var _Object$getOwnPropertyDescriptor__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptor);
var _forEachInstanceProperty__default = /*#__PURE__*/_interopDefault(_forEachInstanceProperty);
var _Object$getOwnPropertyDescriptors__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptors);
var _Object$defineProperties__default = /*#__PURE__*/_interopDefault(_Object$defineProperties);
var _Object$defineProperty__default = /*#__PURE__*/_interopDefault(_Object$defineProperty);
var AccessibleButton__default = /*#__PURE__*/_interopDefault(AccessibleButton);

function ownKeys(e, r) { var t = _Object$keys__default["default"](e); if (_Object$getOwnPropertySymbols__default["default"]) { var o = _Object$getOwnPropertySymbols__default["default"](e); r && (o = _filterInstanceProperty__default["default"](o).call(o, function (r) { return _Object$getOwnPropertyDescriptor__default["default"](e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var _context, _context2; var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? _forEachInstanceProperty__default["default"](_context = ownKeys(Object(t), !0)).call(_context, function (r) { _defineProperty(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors__default["default"] ? _Object$defineProperties__default["default"](e, _Object$getOwnPropertyDescriptors__default["default"](t)) : _forEachInstanceProperty__default["default"](_context2 = ownKeys(Object(t))).call(_context2, function (r) { _Object$defineProperty__default["default"](e, r, _Object$getOwnPropertyDescriptor__default["default"](t, r)); }); } return e; }
const warnIfMissingContent = props => {
  Boolean(props.intlMessage) || Boolean(react$1.Children.count(props.children));
};
const getIconByType = props => {
  switch (props.type) {
    case 'error':
      return icons.ErrorIcon;
    case 'info':
      return icons.InformationIcon;
    case 'warning':
      return icons.WarningIcon;
    default:
      return icons.CheckBoldIcon;
  }
};
const getIconColor = props => {
  switch (props.type) {
    case 'error':
      return designSystem.designTokens.colorError;
    case 'info':
      return designSystem.designTokens.colorInfo;
    case 'warning':
      return designSystem.designTokens.colorWarning;
    case 'success':
      return designSystem.designTokens.fontColorForContentNotificationWhenSuccess;
    default:
      return '';
  }
};
const NotificationTypeIcon = props => {
  const Icon = getIconByType(props);
  return jsxRuntime.jsx("div", {
    css: /*#__PURE__*/react.css("display:flex;align-items:center;fill:", getIconColor(props), ";svg{margin:0 -3px;}" + ("" ), "" ),
    children: jsxRuntime.jsx(Icon, {})
  });
};
NotificationTypeIcon.propTypes = {};
NotificationTypeIcon.displayName = 'NotificationTypeIcon';
const NotificationCloseIcon = props => jsxRuntime.jsx(AccessibleButton__default["default"], {
  label: "Remove",
  onClick: props.onRemove,
  css: /*#__PURE__*/react.css("display:flex;align-items:center;fill:", designSystem.designTokens.colorSolid, ";&:hover{fill:", designSystem.designTokens.colorNeutral40, ";}" + ("" ), "" ),
  children: jsxRuntime.jsx(icons.CloseBoldIcon, {
    size: "medium"
  })
});
NotificationCloseIcon.displayName = 'NotificationCloseIcon';
const getContentBorderColor = props => {
  switch (props.type) {
    case 'error':
      return designSystem.designTokens.colorError85;
    case 'info':
      return designSystem.designTokens.colorInfo85;
    case 'warning':
      return designSystem.designTokens.colorWarning85;
    case 'success':
      return designSystem.designTokens.borderColorForContentNotificationWhenSuccess;
    default:
      return '';
  }
};
const getContainerBackgroundColor = props => {
  switch (props.type) {
    case 'error':
      return designSystem.designTokens.colorError95;
    case 'info':
      return designSystem.designTokens.colorInfo95;
    case 'warning':
      return designSystem.designTokens.colorWarning95;
    case 'success':
      return designSystem.designTokens.backgroundColorForContentNotificationWhenSuccess;
    default:
      return '';
  }
};
const ContentNotification = props => {
  warnIfMissingContent(props);
  return jsxRuntime.jsxs("div", _objectSpread(_objectSpread({}, utils.filterDataAttributes(props)), {}, {
    css: /*#__PURE__*/react.css("display:flex;align-items:stretch;text-align:left;word-break:break-word;hyphens:auto;font-size:", designSystem.designTokens.fontSize30, ";color:", designSystem.designTokens.colorSolid, ";font-family:inherit;background-color:", getContainerBackgroundColor(props), ";border-radius:", designSystem.designTokens.borderRadius4, ";border-width:1px;border-style:solid;border-color:", getContentBorderColor(props), ";padding:", designSystem.designTokens.spacing20, " ", designSystem.designTokens.spacing30, ";" + ("" ), "" ),
    children: [jsxRuntime.jsx(NotificationTypeIcon, {
      type: props.type
    }), jsxRuntime.jsx("div", {
      css: /*#__PURE__*/react.css("flex-grow:1;display:flex;align-items:center;padding:0 ", designSystem.designTokens.spacing30, " 0 ", designSystem.designTokens.spacing20, ";" + ("" ), "" ),
      children: props.intlMessage ? jsxRuntime.jsx("div", {
        children: jsxRuntime.jsx(reactIntl.FormattedMessage, _objectSpread({}, props.intlMessage))
      }) : props.children
    }), props.onRemove && jsxRuntime.jsx(NotificationCloseIcon, {
      onRemove: props.onRemove
    })]
  }));
};
ContentNotification.propTypes = {};
ContentNotification.displayName = 'ContentNotification';
var ContentNotification$1 = ContentNotification;

// NOTE: This string will be replaced on build time with the package version.
var version = "18.3.0";

exports.ContentNotification = ContentNotification$1;
exports.version = version;
