// @commercetools-uikit/stamp/dist/commercetools-uikit-stamp.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _Object$keys = require('@babel/runtime-corejs3/core-js-stable/object/keys');
var _Object$getOwnPropertySymbols = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
var _filterInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/filter');
var _Object$getOwnPropertyDescriptor = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
var _forEachInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
var _Object$getOwnPropertyDescriptors = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
var _Object$defineProperties = require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
var _Object$defineProperty = require('@babel/runtime-corejs3/core-js-stable/object/define-property');
var _defineProperty = require('@babel/runtime-corejs3/helpers/defineProperty');
require('prop-types');
var react = require('react');
var react$1 = require('@emotion/react');
var designSystem = require('@commercetools-uikit/design-system');
var Text = require('@commercetools-uikit/text');
var SpacingsInline = require('@commercetools-uikit/spacings-inline');
var utils = require('@commercetools-uikit/utils');
var jsxRuntime = require('@emotion/react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _Object$keys__default = /*#__PURE__*/_interopDefault(_Object$keys);
var _Object$getOwnPropertySymbols__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertySymbols);
var _filterInstanceProperty__default = /*#__PURE__*/_interopDefault(_filterInstanceProperty);
var _Object$getOwnPropertyDescriptor__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptor);
var _forEachInstanceProperty__default = /*#__PURE__*/_interopDefault(_forEachInstanceProperty);
var _Object$getOwnPropertyDescriptors__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptors);
var _Object$defineProperties__default = /*#__PURE__*/_interopDefault(_Object$defineProperties);
var _Object$defineProperty__default = /*#__PURE__*/_interopDefault(_Object$defineProperty);
var Text__default = /*#__PURE__*/_interopDefault(Text);
var SpacingsInline__default = /*#__PURE__*/_interopDefault(SpacingsInline);

function ownKeys(e, r) { var t = _Object$keys__default["default"](e); if (_Object$getOwnPropertySymbols__default["default"]) { var o = _Object$getOwnPropertySymbols__default["default"](e); r && (o = _filterInstanceProperty__default["default"](o).call(o, function (r) { return _Object$getOwnPropertyDescriptor__default["default"](e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var _context, _context2; var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? _forEachInstanceProperty__default["default"](_context = ownKeys(Object(t), !0)).call(_context, function (r) { _defineProperty(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors__default["default"] ? _Object$defineProperties__default["default"](e, _Object$getOwnPropertyDescriptors__default["default"](t)) : _forEachInstanceProperty__default["default"](_context2 = ownKeys(Object(t))).call(_context2, function (r) { _Object$defineProperty__default["default"](e, r, _Object$getOwnPropertyDescriptor__default["default"](t, r)); }); } return e; }
const tonesPropsMap = themedValue => ({
  critical: {
    styles: {
      backgroundColor: designSystem.designTokens.colorError95,
      borderColor: designSystem.designTokens.colorError85,
      color: designSystem.designTokens.colorError40
    },
    iconColor: 'error'
  },
  warning: {
    styles: {
      backgroundColor: designSystem.designTokens.colorWarning95,
      borderColor: designSystem.designTokens.colorWarning85,
      color: designSystem.designTokens.colorWarning40
    },
    iconColor: 'warning'
  },
  positive: {
    styles: {
      backgroundColor: designSystem.designTokens.backgroundColorForStampAsPositive,
      borderColor: designSystem.designTokens.colorPrimary85,
      color: designSystem.designTokens.fontColorForStampAsPositive
    },
    iconColor: themedValue('primary', 'success')
  },
  information: {
    styles: {
      backgroundColor: designSystem.designTokens.colorInfo95,
      borderColor: designSystem.designTokens.colorInfo85,
      color: designSystem.designTokens.colorInfo40
    },
    iconColor: 'info'
  },
  primary: {
    styles: {
      backgroundColor: designSystem.designTokens.backgroundColorForStampAsPrimary,
      borderColor: designSystem.designTokens.colorPrimary85,
      color: designSystem.designTokens.colorPrimary25
    },
    iconColor: 'primary40'
  },
  secondary: {
    styles: {
      backgroundColor: designSystem.designTokens.colorNeutral95,
      borderColor: designSystem.designTokens.colorNeutral85,
      color: designSystem.designTokens.colorNeutral40
    },
    iconColor: 'neutral60'
  }
});
const getPaddingStyle = props => {
  if (props.isCondensed) return /*#__PURE__*/react$1.css("padding:", designSystem.designTokens.spacing05, " ", designSystem.designTokens.spacing10, ";" + ("" ), "" );
  return /*#__PURE__*/react$1.css("padding:", designSystem.designTokens.spacing10, " ", designSystem.designTokens.spacing25, ";" + ("" ), "" );
};
const getIconColor = (props, overrideTextColor, themedValue) => {
  if (!overrideTextColor) {
    return 'inherit';
  }
  const toneProps = props.tone && tonesPropsMap(themedValue)[props.tone];
  return toneProps ? toneProps.iconColor : '';
};
const getToneStyles = (props, themedValue) => {
  if (!props.tone || !tonesPropsMap(themedValue)[props.tone]) {
    return /*#__PURE__*/react$1.css("" , "" );
  }
  const toneProps = tonesPropsMap(themedValue)[props.tone];
  return /*#__PURE__*/react$1.css("background-color:", toneProps.styles.backgroundColor, ";&,& *{color:", props.overrideTextColor ? toneProps.styles.color : 'inherit', ";}" + ("" ), "" );
};
const getStampStyles = props => {
  return /*#__PURE__*/react$1.css("color:", props.overrideTextColor ? 'inherit' : designSystem.designTokens.colorSolid, ";display:inline-block;border-radius:", props.isCondensed ? designSystem.designTokens.borderRadius2 : designSystem.designTokens.borderRadius4, ";" + ("" ), "" );
};
const Stamp = props => {
  const _useTheme = designSystem.useTheme(),
    themedValue = _useTheme.themedValue;
  const Icon = props.icon && /*#__PURE__*/react.cloneElement(props.icon, {
    size: 'medium',
    color: getIconColor(props, true, themedValue)
  });
  utils.useWarnDeprecatedProp(!Boolean(props.children), 'children', 'Stamp', 'Please use `label` and `icon` properties instead.');
  const StampLabel = _ref => {
    let children = _ref.children;
    return props.isCondensed ? jsxRuntime.jsx(Text__default["default"].Caption, {
      tone: "inherit",
      fontWeight: "medium",
      children: children
    }) : jsxRuntime.jsx(Text__default["default"].Detail, {
      tone: "inherit",
      fontWeight: "medium",
      children: children
    });
  };
  return jsxRuntime.jsx("div", {
    css: [getStampStyles(_objectSpread(_objectSpread({}, props), {}, {
      overrideTextColor: true
    })), getToneStyles(_objectSpread(_objectSpread({}, props), {}, {
      overrideTextColor: true
    }), themedValue), getPaddingStyle(props), "" , "" ],
    children: jsxRuntime.jsxs(SpacingsInline__default["default"], {
      alignItems: "center",
      scale: props.isCondensed ? 'xs' : 's',
      children: [Icon, props.label ? jsxRuntime.jsx(StampLabel, {
        children: props.label
      }) : props.children]
    })
  });
};
Stamp.propTypes = {};
const defaultProps = {
  isCondensed: false,
  tone: 'information'
};
Stamp.displayName = 'Stamp';
Stamp.defaultProps = defaultProps;
var Stamp$1 = Stamp;

// NOTE: This string will be replaced on build time with the package version.
var version = "18.3.0";

exports["default"] = Stamp$1;
exports.version = version;
