// @commercetools-uikit/primary-button/dist/commercetools-uikit-primary-button.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _Object$keys = require('@babel/runtime-corejs3/core-js-stable/object/keys');
var _Object$getOwnPropertySymbols = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
var _filterInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/filter');
var _Object$getOwnPropertyDescriptor = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
var _forEachInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
var _Object$getOwnPropertyDescriptors = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
var _Object$defineProperties = require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
var _Object$defineProperty = require('@babel/runtime-corejs3/core-js-stable/object/define-property');
var _defineProperty = require('@babel/runtime-corejs3/helpers/defineProperty');
require('prop-types');
var react$1 = require('react');
var omit = require('lodash/omit');
var react = require('@emotion/react');
var Inline = require('@commercetools-uikit/spacings-inline');
var designSystem = require('@commercetools-uikit/design-system');
var utils = require('@commercetools-uikit/utils');
var AccessibleButton = require('@commercetools-uikit/accessible-button');
var jsxRuntime = require('@emotion/react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _Object$keys__default = /*#__PURE__*/_interopDefault(_Object$keys);
var _Object$getOwnPropertySymbols__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertySymbols);
var _filterInstanceProperty__default = /*#__PURE__*/_interopDefault(_filterInstanceProperty);
var _Object$getOwnPropertyDescriptor__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptor);
var _forEachInstanceProperty__default = /*#__PURE__*/_interopDefault(_forEachInstanceProperty);
var _Object$getOwnPropertyDescriptors__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptors);
var _Object$defineProperties__default = /*#__PURE__*/_interopDefault(_Object$defineProperties);
var _Object$defineProperty__default = /*#__PURE__*/_interopDefault(_Object$defineProperty);
var omit__default = /*#__PURE__*/_interopDefault(omit);
var Inline__default = /*#__PURE__*/_interopDefault(Inline);
var AccessibleButton__default = /*#__PURE__*/_interopDefault(AccessibleButton);

const getSizeStyles = size => {
  switch (size) {
    // rendering the same style for now to not introduce a breaking change
    case 'small':
    case 'medium':
      return /*#__PURE__*/react.css("height:", designSystem.designTokens.heightForButtonAsMedium, ";border-radius:", designSystem.designTokens.borderRadiusForButtonAsMedium, ";" + ("" ), "" );
    case 'big':
      return /*#__PURE__*/react.css("height:", designSystem.designTokens.heightForButtonAsBig, ";border-radius:", designSystem.designTokens.borderRadiusForButtonAsBig, ";" + ("" ), "" );
    default:
      return /*#__PURE__*/react.css("" , "" );
  }
};
var _ref = {
  name: "vs1rjk",
  styles: "background-color:#b3003e;&:focus,&:hover{background-color:#cc0047;}"
} ;
var _ref2 = {
  name: "dx7qb5",
  styles: "background-color:#dc630a;&:focus,&:hover{background-color:#e7680d;}"
} ;
const getButtonStyles = (isDisabled, isActive, tone, size) => {
  const baseStyles = /*#__PURE__*/react.css("align-items:center;color:", designSystem.designTokens.colorSurface, ";transition:background-color ", designSystem.designTokens.transitionLinear80Ms, ";padding:", designSystem.designTokens.paddingForButton, ";", getSizeStyles(size), ";" + ("" ), "" );
  // "disabled" takes precendece over "active"
  if (isDisabled) {
    return [baseStyles, /*#__PURE__*/react.css("&,&:active,&:hover{background-color:", designSystem.designTokens.backgroundColorForButtonWhenDisabled, ";color:", designSystem.designTokens.colorNeutral60, ";box-shadow:", designSystem.designTokens.shadow0, ";}" + ("" ), "" )];
  }
  if (isActive) {
    const baseActiveStyles = [baseStyles, /*#__PURE__*/react.css("box-shadow:", designSystem.designTokens.shadow0, ";&:hover,&:focus{box-shadow:", designSystem.designTokens.shadow0, ";}" + ("" ), "" )];
    switch (tone) {
      case 'primary':
        // TODO: This custom colors where itroduced in the Merchant Center redign 2023
        // We need to check with designers if we want to introduce them in our color palette
        // They are 'color-primary' with 10% and 5% black opacity
        return [baseActiveStyles, /*#__PURE__*/react.css("background-color:", designSystem.designTokens.backgroundColorForButtonAsPrimary, ";&:focus,&:hover{background-color:", designSystem.designTokens.backgroundColorForButtonAsPrimaryAsDefaultWhenHovered, ";}" + ("" ), "" )];
      case 'urgent':
        // TODO: This custom colors where itroduced in the Merchant Center redign 2023
        // We need to check with designers if we want to introduce them in our color palette
        // They are 'color-warning' with 10% and 5% black opacity
        return [baseActiveStyles, _ref2];
      case 'critical':
        // TODO: This custom colors where itroduced in the Merchant Center redign 2023
        // We need to check with designers if we want to introduce them in our color palette
        // They are 'color-error' with 10% and 5% black opacity
        return [baseActiveStyles, _ref];
      default:
        return baseActiveStyles;
    }
  }
  const baseDefaultStyles = [baseStyles, /*#__PURE__*/react.css("box-shadow:", designSystem.designTokens.shadow0, ";&:hover,&:focus{box-shadow:", designSystem.designTokens.shadow0, ";}&:active{box-shadow:", designSystem.designTokens.shadow0, ";}" + ("" ), "" )];
  switch (tone) {
    case 'primary':
      // TODO: This custom colors where itroduced in the Merchant Center redign 2023
      // We need to check with designers if we want to introduce them in our color palette
      // They are 'color-primary' with 10% and 5% black opacity
      return [baseDefaultStyles, /*#__PURE__*/react.css("background-color:", designSystem.designTokens.colorPrimary, ";&:focus,&:hover{background-color:", designSystem.designTokens.backgroundColorForButtonAsPrimaryAsDefaultWhenHovered, ";}&:active{background-color:", designSystem.designTokens.backgroundColorForButtonAsPrimaryAsDefaultWhenActive, ";}" + ("" ), "" )];
    case 'urgent':
      // TODO: This custom colors where itroduced in the Merchant Center redign 2023
      // We need to check with designers if we want to introduce them in our color palette
      // They are 'color-warning' with 10% and 5% black opacity
      return [baseDefaultStyles, /*#__PURE__*/react.css("background-color:", designSystem.designTokens.backgroundColorForButtonAsPrimaryAsUrgent, ";&:focus,&:hover{background-color:#e7680d;}&:active{background-color:#dc630a;}" + ("" ), "" )];
    case 'critical':
      // TODO: These custom colors where itroduced in the Merchant Center redign 2023
      // We need to check with designers if we want to introduce them in our color palette
      // They are 'color-error' with 10% and 5% black opacity
      return [baseDefaultStyles, /*#__PURE__*/react.css("background-color:", designSystem.designTokens.colorError, ";&:focus,&:hover{background-color:#cc0047;}&:active{background-color:#b3003e;}" + ("" ), "" )];
    default:
      return baseDefaultStyles;
  }
};

function ownKeys(e, r) { var t = _Object$keys__default["default"](e); if (_Object$getOwnPropertySymbols__default["default"]) { var o = _Object$getOwnPropertySymbols__default["default"](e); r && (o = _filterInstanceProperty__default["default"](o).call(o, function (r) { return _Object$getOwnPropertyDescriptor__default["default"](e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var _context, _context2; var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? _forEachInstanceProperty__default["default"](_context = ownKeys(Object(t), !0)).call(_context, function (r) { _defineProperty(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors__default["default"] ? _Object$defineProperties__default["default"](e, _Object$getOwnPropertyDescriptors__default["default"](t)) : _forEachInstanceProperty__default["default"](_context2 = ownKeys(Object(t))).call(_context2, function (r) { _Object$defineProperty__default["default"](e, r, _Object$getOwnPropertyDescriptor__default["default"](t, r)); }); } return e; }
const propsToOmit = ['type'];
const defaultProps = {
  type: 'button',
  size: 'big',
  isToggleButton: false,
  tone: 'primary'
};
const PrimaryButton = props => {
  const buttonAttributes = _objectSpread(_objectSpread({
    'data-track-component': 'PrimaryButton'
  }, utils.filterInvalidAttributes(omit__default["default"](props, propsToOmit))), {}, {
    // if there is a divergence between `isDisabled` and `disabled`,
    // we fall back to `isDisabled`
    disabled: props.isDisabled
  });
  utils.useWarning(!Boolean(props.size === 'small'), 'PrimaryButton `small` value for `size` property has been renamed to `medium`. Please update that value when using this component');
  const isActive = Boolean(props.isToggleButton && props.isToggled);
  return jsxRuntime.jsx(AccessibleButton__default["default"], {
    as: props.as,
    type: props.type,
    buttonAttributes: buttonAttributes,
    label: props.label,
    onClick: props.onClick,
    isToggleButton: props.isToggleButton,
    isToggled: props.isToggled,
    isDisabled: props.isDisabled,
    css: getButtonStyles(props.isDisabled, isActive, props.tone, props.size),
    children: jsxRuntime.jsxs(Inline__default["default"], {
      alignItems: "center",
      scale: "xs",
      children: [Boolean(props.iconLeft) && jsxRuntime.jsx("span", {
        css: /*#__PURE__*/react.css("margin:0 ", designSystem.designTokens.spacing10, " 0 0;display:flex;align-items:center;justify-content:center;" + ("" ), "" ),
        children: props.iconLeft && /*#__PURE__*/react$1.cloneElement(props.iconLeft, {
          color: props.isDisabled ? 'neutral60' : 'surface',
          size: props.size === 'big' ? 'big' : 'medium'
        })
      }), jsxRuntime.jsx("span", {
        children: props.label
      })]
    })
  });
};
PrimaryButton.propTypes = {};
PrimaryButton.defaultProps = defaultProps;
PrimaryButton.displayName = 'PrimaryButton';
var PrimaryButton$1 = PrimaryButton;

// NOTE: This string will be replaced on build time with the package version.
var version = "18.3.0";

exports["default"] = PrimaryButton$1;
exports.version = version;
